src/plan_expand_hypertable.c

```diff
@@ -312,6 +316,138 @@ get_chunk_oids(CollectQualCtx *ctx, PlannerInfo *root, RelOptInfo *rel, Hypertab
 		return get_explicit_chunk_oids(ctx, ht);
 }
 
+#if !(PG96 || PG10)
+
+/*
+ * Check whether a group clause uses a function on an open dimension.
+ *
+ * An open dimension can be partitioned using a function expression. In that
+ * case, try to support partitionwise aggregation both on the raw column and the
+ * function expression. For instance, a query could look like
+ *
+ * SELECT time, device, avg(temp)
+ * FROM hypertable
+ * GROUP BY 1, 2;
+ *
+ * or,
+ *
+ * SELECT func(time), device, avg(temp)
+ * FROM hypertable
+ * GROUP BY 1, 2;
+ *
+ * We can support partitionwise aggregation on the first query, but also on the
+ * second if 'func' is the partitioning function on the open "time" dimension.
+ *
+ * Thus, this function will return TRUE, iff either one of these two cases
+ * occur.
+ */
+static bool
+open_dim_uses_partfunc(List **groupexprs, Query *parse, Dimension *dim)
+{
+	if (!IS_OPEN_DIMENSION(dim) || NULL == dim->partitioning)
+		return false;
+
+	/* Lazy set groupexpr */
+	if (*groupexprs == NIL)
+		*groupexprs = get_sortgrouplist_exprs(parse->groupClause, parse->targetList);
+
+	return list_member(*groupexprs, dim->partitioning->partfunc.func_fmgr.fn_expr);
+}
+
+/*
+ * Create partition expressions for a hypertable.
+ *
+ * Build an array of partition expressions where each element is a list of
+ * expression for a particular partitioning dimension.
+ *
+ * The partition expressions are used by, e.g., group_by_has_partkey() to check
+ * whether a GROUP BY clause covers all partitioning dimesions.
+ */
+static List **
+get_hypertable_partexprs(Hypertable *ht, Query *parse, Index varno)
+{
+	int i;
+	List **partexprs;
+	List *groupexprs = NIL;
+
+	Assert(NULL != ht->space);
+
+	partexprs = palloc0(sizeof(List *) * ht->space->num_dimensions);
+
+	for (i = 0; i < ht->space->num_dimensions; i++)
+	{
+		Dimension *dim = &ht->space->dimensions[i];
+		Expr *expr;
+
+		/* Only functions on open dimensions need to be function expressions */
```

Comment:
why doesn't this apply to closed dimensions?

Reply:
I guess it could apply to both closed and open dimensions, but I didn't see an obvious use case for querying a closed dimension by its hash function (i.e., its hash value). That would look something like:

```
SELECT time, _timescaledb_internal.get_partition_hash(device), temp
FROM hypertable
GROUP BY 1, 2;
```

In contrast, it seems feasible to query an open dimension both by the "original" value and the "transformed" value. For instance, like shown in tests, you can store time in floats, but use a "partitioning" function like `to_timestamp()` that turns floats into timestamps. 

Still, I guess I can enable this for closed dimensions too, although it will probably be a bit less efficient since it needs to transform the groupexpr, etc.

Reply:
Added this for closed dimensions too, although not sure of usefulness.